clients/src/test/java/org/apache/kafka/common/network/SSLSelectorTest.java

```diff
@@ -273,4 +320,84 @@ private void sendAndReceive(String node, String requestPrefix, int startIndex, i
+                SSLEngine sslEngine = sslFactory.createSSLEngine(socketChannel.socket().getInetAddress().getHostName(),
+                                socketChannel.socket().getPort());
+                return new TestSSLTransportLayer(id, key, sslEngine, netReadBufSize, netWriteBufSize, appBufSize);
+            }
+
+
+        };
+        this.channelBuilder.configure(sslClientConfigs);
+        this.selector = new Selector(5000, new Metrics(), new MockTime(), "MetricGroup", new LinkedHashMap<String, String>(), channelBuilder);
+    }
+
+    /**
+     * SSLTransportLayer with overrides for packet and application buffer size to test buffer resize
+     * code path. The overridden buffer size starts with a small value and increases in size when the buffer
+     * size is retrieved to handle overflow/underflow, until the actual session buffer size is reached.
+     */
+    private static class TestSSLTransportLayer extends SSLTransportLayer {
+
+        private int netReadBufSizeOverride;
+        private int netWriteBufSizeOverride;
+        private int appBufSizeOverride;
+        private boolean isInitialized;
+
+        public TestSSLTransportLayer(String channelId, SelectionKey key, SSLEngine sslEngine, 
+                int netReadBufSize, int netWriteBufSize, int appBufSize) throws IOException {
+            super(channelId, key, sslEngine);
+            this.netReadBufSizeOverride = netReadBufSize;
+            this.netWriteBufSizeOverride = netWriteBufSize;
+            this.appBufSizeOverride = appBufSize;
+            this.isInitialized = true;
+        }
+        
+        @Override
+        protected int netReadBufferSize() {
+            int size = super.netReadBufferSize();
+            if (!isInitialized) {
+                size = 10;
+            } else if (netReadBufSizeOverride != 0) {
+                if (!netReadBuffer().hasRemaining()) {
+                    netReadBufSizeOverride = Math.min(netReadBufSizeOverride * 2, size);
+                }
+                size = netReadBufSizeOverride;
+            }
+            return size;
+        }
+        
+        @Override
+        protected int netWriteBufferSize() {
+            int size = super.netWriteBufferSize();
+            if (!isInitialized) {
+                size = 10;
+            } else if (netWriteBufSizeOverride != 0) {
+                netWriteBufSizeOverride = Math.min(netWriteBufSizeOverride * 2, size);
+                size = netWriteBufSizeOverride;
+            }
+            return size;
+        }
+
+        @Override
+        protected int applicationBufferSize() {
+            int size = super.applicationBufferSize();
+            if (!isInitialized) {
+                size = 10;
+            } else if (appBufSizeOverride != 0) {
+                appBufSizeOverride = Math.min(appBufSizeOverride * 2, size);
+                size = appBufSizeOverride;
+            }
+            return size;
```

Comment:
It seems like the logic for all the methods is quite similar. Can we not extract a utility method that can be reused?